stan/math/rev/functor/algebra_solver_powell.hpp

```diff
@@ -21,57 +22,77 @@ namespace math {
-                      const Eigen::Matrix<T, Eigen::Dynamic, 1>& y,
-                      const std::vector<double>& dat,
-                      const std::vector<int>& dat_int,
-                      const Eigen::VectorXd& theta_dbl, Fx& fx,
-                      std::ostream* msgs)
-      : vari(theta_dbl(0)),
-        y_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(y.size())),
+  algebra_solver_vari(Fy& fy, const Eigen::Matrix<T, Eigen::Dynamic, 1>& y,
+                      Fx& fx, const Eigen::VectorXd& x)
+      : vari(x(0)),
         y_size_(y.size()),
+        y_val_(y),
+        fy_(fy),
+        y_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(y_size_)),
         x_size_(x.size()),
-        theta_(
-            ChainableStack::instance_->memalloc_.alloc_array<vari*>(x_size_)),
-        Jx_y_(ChainableStack::instance_->memalloc_.alloc_array<double>(
-            x_size_ * y_size_)) {
+        x_val_(x),
+        Jf_x_(fx.get_jacobian(x_val_)),
+        fx_(fx),
+        x_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(x_size_)) {
     using Eigen::Map;
     using Eigen::MatrixXd;
-    for (int i = 0; i < y.size(); ++i) {
+
+    for (int i = 0; i < y_size_; ++i) {
       y_[i] = y(i).vi_;
     }
 
-    theta_[0] = this;
-    for (int i = 1; i < x.size(); ++i) {
-      theta_[i] = new vari(theta_dbl(i), false);
+    x_[0] = this;
+    for (int i = 1; i < x_size_; ++i) {
+      x_[i] = new vari(x(i), false);
     }
-
-    // Compute the Jacobian and store in array, using the
-    // implicit function theorem, i.e. Jx_y = Jf_y / Jf_x
-    using f_y = hybrj_functor_solver<Fs, F, double, double>;
-    Map<MatrixXd>(&Jx_y_[0], x_size_, y_size_)
-        = -mdivide_left(fx.get_jacobian(theta_dbl),
-                        f_y(fs, f, theta_dbl, value_of(y), dat, dat_int, msgs)
-                            .get_jacobian(value_of(y)));
   }
 
   void chain() {
-    for (int j = 0; j < y_size_; j++) {
-      for (int i = 0; i < x_size_; i++) {
-        y_[j]->adj_ += theta_[i]->adj_ * Jx_y_[j * x_size_ + i];
+    using Eigen::Dynamic;
+    using Eigen::Matrix;
+    using Eigen::MatrixXd;
+    using Eigen::VectorXd;
+
+    // Compute (transpose of) specificities with respect to x.
+    VectorXd x_bar_(x_size_);  // TODO: Is this zeroing out memory?
+    for (int i = 0; i < x_size_; ++i) {
+      x_bar_[i] = x_[i]->adj_;
+    }
+
+    // Contract specificities with inverse Jacobian of f with respect to x.
+    VectorXd eta_ = -Jf_x_.transpose().fullPivLu().solve(x_bar_);
+
+    // Contract with Jacobian of f with respect to y using a nested reverse
+    // autodiff pass.
+    {
+      stan::math::nested_rev_autodiff rev;
+      Matrix<var, Eigen::Dynamic, 1> y_nrad_ = y_val_;
+      auto x_nrad_ = stan::math::eval(fy_(y_nrad_));
+      x_nrad_.adj() = eta_;
+      stan::math::grad();
+      for (int j = 0; j < y_size_; j++) {
+        y_[j]->adj_ += y_nrad_.adj()[j];
       }
     }
```

Comment:
@bbbales2 sanity check for me, you can't call grad() in a `chain()` method since that chain method is only called while we've already started going through the callback stack so idt this works. I think this needs to be moved into the constructor storing the adjoints for `y_` and `x_` then updating them in the chain method?

Reply:
No, you can do that. This is also done in the adjoint ODE code. It used to not work for a while, but it is fine by now.

Reply:
Huh neat!